### PR TITLE
Variety of performance improvements

### PR DESCRIPTION
**Pull request by travissanderson-wf**
While profiling I saw a surprising amount of time spent on `new.Rectangle`. When we made r_tree null-safe we decided to make boundingRect non-nullable, which resulted in creating a lot of identical empty Rectangles in some cases. I've switched this to use a shared const Rectangle to represent the uninitialized state. The benchmark results are surprisingly good. We spend less time newing up instances of Rectangles and less time garbage collecting those unused instances.

javascript benchmark  | master | const Rectangle | improvement
-- | -- | -- | --
Insert 100 | 24340.42 | 24954.11 | -2.52%
Insert 1000 | 517026.75 | 493413.47 | 4.57%
Insert 10000 | 11240350.00 | 10192000.00 | 9.33%
Load 100 | 2953.30 | 3150.59 | -6.68%
Load 1000 | 66943.79 | 64563.75 | 3.56%
Load 10000 | 897272.39 | 871911.11 | 2.83%
Remove 5k | 1314433.50 | 1208266.50 | 8.08%
Search/Iterate (using Insert) 100 | 72379.49 | 63770.45 | 11.89%
Search/Iterate (using Insert) 1000 | 309299.95 | 268866.67 | 13.07%
Search/Iterate (using Insert) 10000 | 955705.56 | 797455.44 | 16.56%
Search/Iterate (using Load) 100 | 88016.66 | 77249.92 | 12.23%
Search/Iterate (using Load) 1000 | 310976.14 | 274938.14 | 11.59%
Search/Iterate (using Load) 10000 | 822033.33 | 728433.44 | 11.39%

The difference in the VM is even better in some cases, but pretty similar
VM benchmark | master | const Rectangle | improvement
-- | -- | -- | --
Insert 100 | 7569.75 | 6618.97 | 12.56%
Insert 1000 | 176100.47 | 156262.70 | 11.27%
Insert 10000 | 3841454.33 | 3383032.83 | 11.93%
Load 100 | 1409.82 | 1393.85 | 1.13%
Load 1000 | 48680.43 | 49195.70 | -1.06%
Load 10000 | 706903.56 | 712481.89 | -0.79%
Remove 5k | 357592.89 | 298343.62 | 16.57%
Search/Iterate (using Insert) 100 | 15733.93 | 14652.69 | 6.87%
Search/Iterate (using Insert) 1000 | 68642.47 | 63117.71 | 8.05%
Search/Iterate (using Insert) 10000 | 218887.00 | 196788.16 | 10.10%
Search/Iterate (using Load) 100 | 22216.95 | 21650.92 | 2.55%
Search/Iterate (using Load) 1000 | 79380.16 | 72371.68 | 8.83%
Search/Iterate (using Load) 10000 | 205157.06 | 184301.74 | 10.17%

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

**Review by kerbyseeley-wf — APPROVED** [review]
Wow.  Good find!

File: lib/src/r_tree/node.dart [hunk 1]
```diff
@@ -16,6 +16,8 @@
 
 part of r_tree;
 
+const noMBR = const Rectangle(0, 0, 0, 0);
```
Comment (brettmeyer-wk) [inline lib/src/r_tree/node.dart:19]: `const` is redundant in this context
```suggestion
const noMBR = Rectangle(0, 0, 0, 0);
```

File: lib/src/r_tree/leaf_node.dart [hunk 2]
```diff
@@ -58,6 +58,6 @@ class LeafNode<E> extends Node<E> {
```
Comment (brettmeyer-wk) [inline lib/src/r_tree/leaf_node.dart:58]: 🧵 There are two instances in both `non_leaf_node_test.dart` and `leaf_node_test.dart` I'd update as well.

**Review by kerbyseeley-wf — APPROVED** [review]

File: lib/src/r_tree/leaf_node.dart [hunk 3]
```diff
@@ -58,6 +58,6 @@ class LeafNode<E> extends Node<E> {
 
   clearChildren() {
     _items.clear();
-    _minimumBoundingRect = Rectangle(0, 0, 0, 0);
+    _minimumBoundingRect = null;
```
Comment (brettmeyer-wk) [inline lib/src/r_tree/leaf_node.dart:61]: Why the change to `null`?

File: lib/src/r_tree/node.dart [hunk 4]
```diff
@@ -28,14 +30,16 @@ abstract class Node<E> extends RTreeContributor {
   /// Parent node of this node, or null if this is the root node
   Node<E>? parent;
 
-  Rectangle _minimumBoundingRect = Rectangle(0, 0, 0, 0);
+  Rectangle? _minimumBoundingRect = null;
```
Comment (brettmeyer-wk) [inline lib/src/r_tree/node.dart:33]:
```suggestion
  Rectangle? _minimumBoundingRect;
```

File: lib/src/r_tree/leaf_node.dart (hunk 3 above)
Comment (travissanderson-wf) [reply lib/src/r_tree/leaf_node.dart:61]: I was talking to @danielway-wk about it and he suggested null is a better representation, which is what we used pre-null-safety. I decided to try changing it in a way that didn't ripple nulls all the way out and was surprised that it actually made performance even better. I haven't dug into why yet so I haven't posted the results, but they were significant. Daniel reproduced the same gain on his machine for my sanity.

Comment (danielway-wk) [reply lib/src/r_tree/leaf_node.dart:61]: > ...he suggested null is a better representation

To elaborate on this point: I would argue a node without children has no bounding rectangle. `null` seems like a clearer way to indicate that then `Rectangle(0, 0, 0, 0)` in my opinion.

Comment (danielway-wk) [reply lib/src/r_tree/leaf_node.dart:61]: > ...he suggested null is a better representation

To elaborate on this point: I would argue a node without children has no bounding rectangle. `null` seems like a clearer way to indicate that then `Rectangle(0, 0, 0, 0)` in my opinion.

Comment (brettmeyer-wk) [reply lib/src/r_tree/leaf_node.dart:61]: I figured that was the case. Of course you could always do `Rectangle(-1, -1, -1, -1)` 😉

**Review by brettmeyer-wk — DISMISSED** [review]

**Review by kerbyseeley-wf — APPROVED** [review]

**Comment by travissanderson-wf** [comment]
Ran some final numbers while this is being QA'd by @kentiverson-wf 

  | master | branch | vs master
-- | -- | -- | --
Insert 100 | 20879.40 | 13618.28 | 34.78%
Insert 1000 | 441479.87 | 370616.72 | 16.05%
Insert 10000 | 9552949.83 | 7762783.50 | 18.74%
Load 100 | 2587.08 | 2934.69 | -13.44%
Load 1000 | 62039.88 | 65160.08 | -5.03%
Load 10000 | 840933.34 | 864966.55 | -2.86%
Remove 5k | 1167183.33 | 616483.33 | 47.18%
Search/Iterate (using Insert) 100 | 55959.76 | 27696.30 | 50.51%
Search/Iterate (using Insert) 1000 | 249051.85 | 108691.23 | 56.36%
Search/Iterate (using Insert) 10000 | 770311.11 | 308204.86 | 59.99%
Search/Iterate (using Load) 100 | 69369.50 | 36867.85 | 46.85%
Search/Iterate (using Load) 1000 | 248396.26 | 128856.25 | 48.12%
Search/Iterate (using Load) 10000 | 686255.56 | 319204.81 | 53.49%

**Comment by travissanderson-wf** [comment]
Final numbers for the two PRs combined. I pulled the other PR into this one to trigger CI and format the code.

  | master | [less new rect ](https://github.com/Workiva/r_tree/pull/58)| vs master | [ronbb](https://github.com/Workiva/r_tree/pull/55) | vs master | integrated | vs master
-- | -- | -- | -- | -- | -- | -- | --
Insert 100 | 20879.40 | 13618.28 | 34.78% | 21543.90 | -3.18% | 14070.62 | 32.61%
Insert 1000 | 441479.87 | 370616.72 | 16.05% | 460039.93 | -4.20% | 380177.83 | 13.89%
Insert 10000 | 9552949.83 | 7762783.50 | 18.74% | 9894083.50 | -3.57% | 7937750.17 | 16.91%
Load 100 | 2587.08 | 2934.69 | -13.44% | 2151.70 | 16.83% | 2454.04 | 5.14%
Load 1000 | 62039.88 | 65160.08 | -5.03% | 40940.56 | 34.01% | 43835.25 | 29.34%
Load 10000 | 840933.34 | 864966.55 | -2.86% | 517308.33 | 38.48% | 545866.83 | 35.09%
Remove 5k | 1167183.33 | 616483.33 | 47.18% | 677688.66 | 41.94% | 415553.33 | 64.40%
Search/Iterate (using Insert) 100 | 55959.76 | 27696.30 | 50.51% | 33572.67 | 40.01% | 23375.80 | 58.23%
Search/Iterate (using Insert) 1000 | 249051.85 | 108691.23 | 56.36% | 132062.50 | 46.97% | 88768.09 | 64.36%
Search/Iterate (using Insert) 10000 | 770311.11 | 308204.86 | 59.99% | 395050.00 | 48.72% | 252308.42 | 67.25%
Search/Iterate (using Load) 100 | 69369.50 | 36867.85 | 46.85% | 41045.45 | 40.83% | 29914.37 | 56.88%
Search/Iterate (using Load) 1000 | 248396.26 | 128856.25 | 48.12% | 144276.19 | 41.92% | 104539.98 | 57.91%
Search/Iterate (using Load) 10000 | 686255.56 | 319204.81 | 53.49% | 383944.45 | 44.05% | 267487.50 | 61.02%

File: lib/src/r_tree/quickselect.dart [hunk 5]
```diff
@@ -3,9 +3,8 @@ part of r_tree;
 
 // sort an array so that items come in groups of n unsorted items, with groups sorted between each other;
 // combines selection algorithm with binary divide & conquer approach
-multiSelect<E>(List<E> arr, int left, int right, int n, num Function(E) getter) {
+multiSelect<E>(List<E> arr, int left, int right, int n, int Function(E a, E b) compare) {
```
Comment (brettmeyer-wk) [inline lib/src/r_tree/quickselect.dart:6]: #nit
```suggestion
void multiSelect<E>(List<E> arr, int left, int right, int n, int Function(E a, E b) compare) {
```

**Review by brettmeyer-wk — APPROVED** [review]

**Comment by kentiverson-wf** [comment]
QA +1
- [x] Created a DPC branch to consume these changes and ran it through skynet, all CI tests [passed](https://wf-skynet-hrd.appspot.com/apps/test/smithy/5148351/1)
- [x] CI passing

File: lib/src/r_tree/quickselect.dart (hunk 5 above)
Comment (travissanderson-wf) [reply lib/src/r_tree/quickselect.dart:6]: we should probably add some stronger analysis options in a future PR

**Comment by travissanderson-wf** [comment]
@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

File: lib/src/r_tree/quickselect.dart (hunk 5 above)
Comment (brettmeyer-wk) [reply lib/src/r_tree/quickselect.dart:6]: #allthelints same with datatables
